src/argument_parser.h

```diff
@@ -0,0 +1,169 @@
+#include "bech32m_exception.h"
+#include <functional>
+#include <iostream>
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+#include <sstream>
+
+#ifndef ARGUMENT_PARSER
+#define ARGUMENT_PARSER
+enum class Mode { Encode, Decode };
+enum class DataFormat { Bin, Hex, Base64, Bech32m };
+enum class Output {
+    Stdout,
+    File,
+};
+enum class Input { Stdin, File, Argument };
+
+const char *const BIN_STRING = "bin";
+const char *const HEX_STRING = "hex";
+const char *const BASE64_STRING = "base64";
+const char *const BECH32M_STRING = "bech32m";
+
+struct ProgramArgs {
+    Mode mode = Mode::Encode;
+
+    Input input_type = Input::Stdin;
+    std::string input_file;
+    std::string input_text;
+
+    // used in Encode mode
+    DataFormat input_format = DataFormat::Hex;
+
+    Output output_type = Output::Stdout;
+    std::string output_file;
+    // used in Decode mode
+    DataFormat output_format = DataFormat::Bech32m;
+
+    bool print_help = false;
+
+    // flags to prevent setting both formats or one format repeatably
+    bool oformat_set = false;
+    bool iformat_set = false;
+
+    // hrp handling
+    bool allow_empty_hrp = false;
+    std::string hrp;
+
+    bool trim_trailing_zeros = false;
+};
+
+class Argument {
+  public:
+    using HandlerType = std::function<void(ProgramArgs &, std::string)>;
+
+  private:
+    // set of valid parameters for the argument
+    std::set<std::string> valid_params;
+    // set of valid names for the argument
+    std::string name;
+    // description of the argument
+    std::string description = "No description provided.";
+    // function pointer
+    HandlerType handler;
+
+
+    bool param = false;
+    bool variable_parameter = false;
+
+  public:
+    bool has_param() const { return param; }
+    bool has_variable_param() const { return variable_parameter; }
+
+    Argument set_variable_param() {
+        param = true;
+        variable_parameter = true;
+        return *this;
```

Comment:
As per this [link](https://stackoverflow.com/a/26614794) the return value could probably be a reference to Argument& instead of just Argument